crates/autopilot/src/run_loop.rs

```diff
@@ -98,6 +91,47 @@ impl RunLoop {
         }
     }
 
+    /// Waits until it's the correct time to start the next auction.
+    async fn sleep_until_next_auction(&self) {
+        if let RunLoopMode::SyncToBlockchain = self.synchronization {
+            // The time on the blockchain is discrete because things only happen inside
+            // blocks which get published in set time intervals.
+            // Therefore any critical work should ideally start right after a new block
+            // was seen in order to improve the chances that your work is done **before**
+            // the next block gets published so you can submit a transaction in time.
+            //
+            // Intuitively that means just waiting for the next block and then starting to
+            // work. However, in our context it's important to know whether the transaction
+            // we tried to submit actually made it on-chain (i.e. into the next block).
+            //
+            // This would effectively mean that we'd always have almost 1 full block of
+            // downtime between iterations:
+            // 1. wait for new block
+            // 2. do work
+            // 3. submit tx
+            // 4. wait for new block to see if our tx made it in
+            // 5. jump to 1 (which now does nothing for an entire block interval)
+            //
+            // To avoid this we have another condition we await. It's a signal from the
+            // transaction submission logic that tells us to start right away and **not**
+            // wait for the next block.
+            let await_new_block = ethrpc::block_stream::next_block(self.eth.current_block());
+            pin_mut!(await_new_block);
+            let await_notify = self.run_loop_start_notifier.notified();
+            pin_mut!(await_notify);
+            futures::future::select(await_notify, await_new_block).await;
+
+            let block_number = self.eth.current_block().borrow().number;
+            if let Err(err) = self.solvable_orders_cache.update(block_number).await {
```

Comment:
I think we also need to make sure our COMBO condition is satisfied before building auction here. But this could also be an optimization over current implementation, not necessarily blocking.

Reply:
Why should the COMBO condition from the transaction detection discussion be relevant here?
But regardless the tx detection code await the COMBO condition and if we reached the deadline it informs this code to NOT wait for the next block and instead start right away.
This is how it's supposed to be from my understanding or am I missing something?

Reply:
> Why should the COMBO condition from the transaction detection discussion be relevant here?

So that auction is built on top of executed trades from the newest block. DB needs Trade events indexed so it can update the state of orders and return only OPEN orders to SolvableOrdersCache. Then we can also drop inflight orders. But as I said, I am not sure this would be a requirement for this PR.

Reply:
AFAIU unless we run into the issue discussed [here](https://github.com/cowprotocol/services/pull/2900#discussion_r1719591771) we should be okay. Will think more about avoiding that issue.